src/python/pants/engine/goal.py

```diff
@@ -1,33 +1,19 @@
 # Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
 
-from abc import ABCMeta, abstractmethod
+from abc import abstractmethod
 from contextlib import contextmanager
 from dataclasses import dataclass
+from typing import ClassVar, Type
 
 from pants.cache.cache_setup import CacheSetup
 from pants.option.optionable import Optionable
 from pants.option.scope import ScopeInfo
 from pants.subsystem.subsystem_client_mixin import SubsystemClientMixin
-from pants.util.memo import memoized_classproperty
 from pants.util.meta import classproperty
 
 
-@dataclass(frozen=True)  # type: ignore[misc] # tracked by https://github.com/python/mypy/issues/5374, which they put as high priority.
-class Goal(metaclass=ABCMeta):
-  """The named product of a `@console_rule`.
-
-  This abstract class should be subclassed and given a `Goal.name` that it will be referred to by
-  when invoked from the command line. The `Goal.name` also acts as the options_scope for the `Goal`.
-
-  Since `@console_rules` always run in order to produce side effects (generally: console output), they
-  are not cacheable, and the `Goal` product of a `@console_rule` contains only a exit_code value to
-  indicate whether the rule exited cleanly.
-
-  Options values for a Goal can be retrived by declaring a dependency on the relevant `Goal.Options`
-  class.
-  """
-  exit_code: int
+class GoalSubsystem(SubsystemClientMixin, Optionable):
 
```

Comment:
This could do with a doc comment. Something about the subsystem for a goal, which is where options are registered, and the name being the root scope for options in this subsystem.